### PR TITLE
domain/infosync: Adding topology syncer

### PR DESCRIPTION
**Pull request by mapleFU**
Signed-off-by: mapleFU <1506118561@qq.com>

### What problem does this PR solve? 

TiDB-Dashboard require tidb's binary-path and time to live. This pr add a topology syncer in `domain/infosync`. This pr add two path in etcd:
* `/topology/tidb/{host}:{ip}/tidb` store json for tidb topology.
*  `/topology/tidb/{host}:{ip}/ttl` store last updated time.

### What is changed and how it works?

changed `domain/domain.go` and `domain/infosync/info.go`

### Check List 

Tests 

 - No code

Code changes

 - Has exported function/method change
 - Has exported variable/fields change
 - Has persistent data change

Side effects

 - Possible performance regression

Related changes

Release note

 - Write release note for bug-fix or new feature.

### CODE REVIEW

File: domain/infosync/info.go [hunk 1]
```diff
@@ -78,6 +86,8 @@ type ServerInfo struct {
 	Lease          string `json:"lease"`
 	BinlogStatus   string `json:"binlog_status"`
 	StartTimestamp int64  `json:"start_timestamp"`
+
+	Host string `json:"-"`
```
Comment (lonng) [inline domain/infosync/info.go]: Do we need this field? (can we use the `IP` field instead of adding a new field)

File: domain/infosync/info.go [hunk 2]
```diff
@@ -54,17 +55,24 @@ const (
 	InfoSessionTTL = 10 * 60
 	// ReportInterval is interval of infoSyncerKeeper reporting min startTS.
 	ReportInterval = 30 * time.Second
+	// TopologyInformationPath means etcd path for storing topology info.
+	TopologyInformationPath = "/topology/tidb/"
+	// TopologyTimeToLive is ttl for topology.
+	TopologyTimeToLive = 60 * time.Second
```
Comment (breeswish) [inline domain/infosync/info.go]: 1 minute is too long. How about 30 seconds TTL + 15 seconds refresh

File: domain/infosync/info.go [hunk 3]
```diff
@@ -410,6 +474,7 @@ func getServerInfo(id string) *ServerInfo {
 		Lease:          cfg.Lease,
 		BinlogStatus:   binloginfo.GetStatus().String(),
 		StartTimestamp: time.Now().Unix(),
+		Host:           cfg.Host,
```
Comment (breeswish) [inline domain/infosync/info.go]: Why do we need to change this?

File: domain/infosync/info.go (hunk 2 above)
Comment (mapleFU) [reply domain/infosync/info.go]: I think both are OK...

File: domain/infosync/info.go [hunk 4]
```diff
@@ -58,7 +58,7 @@ const (
 	// TopologyInformationPath means etcd path for storing topology info.
 	TopologyInformationPath = "/topology/tidb/"
 	// TopologyTimeToLive is ttl for topology.
-	TopologyTimeToLive = 60 * time.Second
+	TopologyTimeToLive = 45
```
Comment (breeswish) [inline domain/infosync/info.go]: Please comment why there is no `* time.Second` as above. Please also follow the name convention `TTL` as above.

Comment (mapleFU) [reply domain/infosync/info.go]: Done

File: domain/infosync/info.go (hunk 2 above)
Comment (mapleFU) [reply domain/infosync/info.go]: I change it to 45 secs now.

File: domain/infosync/info.go [hunk 5]
```diff
@@ -271,6 +279,44 @@ func (is *InfoSyncer) RemoveServerInfo() {
...
+// RemoveServerInfo stores the topology of tidb to etcd.
+// Note: this part no ttl exists.
+func (is *InfoSyncer) storeTopologyInfo(ctx context.Context) error {
+	if is.etcdCli == nil {
+		return nil
+	}
+	// maybe it's not proper here.
+	s, err := os.Executable()
+	if err != nil {
+		return err
```
Comment (breeswish) [inline domain/infosync/info.go]: Return error is not proper. This will completely miss the info.

File: domain/infosync/info.go [hunk 6]
```diff
@@ -271,6 +279,44 @@ func (is *InfoSyncer) RemoveServerInfo() {
 	}
 }
 
+type topologyInfo struct {
+	ServerVersionInfo
+	StatusPort uint
```
Comment (breeswish) [inline domain/infosync/info.go]: Why there is no json tags? Will it result in inconsistent field naming convension?

File: domain/infosync/info.go [hunk 7]
```diff
@@ -359,9 +405,31 @@ func (is *InfoSyncer) newSessionAndStoreServerInfo(ctx context.Context, retryCnt
...
+	if err != nil {
+		return err
+	}
+	is.topologySession = session
+	err = is.storeTopologyInfo(ctx)
 	return err
 }
 
+// RefreshTopology refreshes etcd topology with ttl stored in "/topology/tidb/ip:port/ttl".
+func (is *InfoSyncer) RefreshTopology(ctx context.Context) error {
```
Comment (breeswish) [inline domain/infosync/info.go]: Why `/topology/tidb/ip:port/tidb` is not refreshed? What if user deletes it?

Comment (mapleFU) [reply domain/infosync/info.go]: You're right, I think we have to ways to solve this:
* Refresh `tidb` in `RefreshTopology`.
* Watch and using `IsModify()` to check `tidb` in `topologySyncerKeeper`  
I think the second is better, but maybe `Restart` will be conflict with it.

File: domain/infosync/info.go [hunk 8]
```diff
@@ -271,6 +279,44 @@ func (is *InfoSyncer) RemoveServerInfo() {
...
+	}
+
+	infoBuf, err := json.Marshal(topologyInfo)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	str := string(hack.String(infoBuf))
+	key := fmt.Sprintf("%s/%s:%v/tidb", TopologyInformationPath, is.info.IP, is.info.Port)
+	// Note: no lease is required here.
+	err = util.PutKVToEtcd(ctx, is.etcdCli, keyOpDefaultRetryCnt, key, str)
```
Comment (crazycs520) [inline domain/infosync/info.go]: with lease will be better?
If tidb panic unexpected, the info won't be deleted.

Comment (mapleFU) [reply domain/infosync/info.go]: If it's not ttl session, we can just leave it here.

Comment (crazycs520) [reply domain/infosync/info.go]: I mean, if TiDB panic, the KV will never be deleted, except user manual delete.

Comment (mapleFU) [reply domain/infosync/info.go]: (Sure, It was designed like this, user can delete non-ttl key in tidb-dashboard)

File: domain/infosync/info.go (hunk 6 above)
Comment (mapleFU) [reply domain/infosync/info.go]: No, I just fixed it.

File: domain/infosync/info.go [hunk 9]
```diff
@@ -358,8 +430,34 @@ func (is *InfoSyncer) newSessionAndStoreServerInfo(ctx context.Context, retryCnt
...
-	return err
+	return is.storeServerInfo(ctx)
+}
+
+// newTopologySessionAndStoreServerInfo creates a new etcd session and stores server info to etcd.
+func (is *InfoSyncer) newTopologySessionAndStoreServerInfo(ctx context.Context, retryCnt int) error {
+	if is.etcdCli == nil {
+		return nil
+	}
+	logPrefix := fmt.Sprintf("[topology-syncer] %s", is.serverInfoPath)
```
Comment (lonng) [inline domain/infosync/info.go]: The `logPrefix` of topology session should not use the `is.serverInfoPath`, which may cause confusion.

File: domain/infosync/info.go [hunk 10]
```diff
@@ -358,8 +430,34 @@ func (is *InfoSyncer) newSessionAndStoreServerInfo(ctx context.Context, retryCnt
...
+	logPrefix = fmt.Sprintf("[topology-syncer] %s", is.serverInfoPath)
```
Comment (lonng) [inline domain/infosync/info.go]: ditto

**Comment by mapleFU** [comment]
@lonng PTAL again

**Comment by lonng** [comment]
Rest LGTM, but I think it's better to change the path `/topology/tidb/{host}:{ip}/tidb` to `/topology/tidb/{host}:{ip}/info`

**Comment by mapleFU** [comment]
@lonng changed.

File: domain/domain.go [hunk 11]
```diff
@@ -456,6 +456,30 @@ func (do *Domain) infoSyncerKeeper() {
 	}
 }
 
+func (do *Domain) topologySyncerKeeper() {
+	defer do.wg.Done()
+	ticker := time.NewTicker(infosync.TopologyTimeToRefresh)
```
Comment (breeswish) [inline domain/domain.go]: Do we need to recover in domain like infoSyncer?

```
defer recoverInDomain("infoSyncerKeeper", false)
```

File: domain/infosync/info.go [hunk 12]
```diff
@@ -358,8 +422,33 @@ func (is *InfoSyncer) newSessionAndStoreServerInfo(ctx context.Context, retryCnt
...
+	if err != nil {
+		return err
+	}
+
+	is.topologySession = session
+	return is.StoreTopologyInfo(ctx)
+}
+
+// refreshTopology refreshes etcd topology with ttl stored in "/topology/tidb/ip:port/ttl".
+func (is *InfoSyncer) refreshTopology(ctx context.Context) error {
```
Comment (breeswish) [inline domain/infosync/info.go]: The name is confusing. You should call it `updateTopologyAliveness`

File: domain/infosync/info.go [hunk 13]
```diff
@@ -358,8 +422,33 @@ func (is *InfoSyncer) newSessionAndStoreServerInfo(ctx context.Context, retryCnt
...
+	if is.etcdCli == nil {
+		return nil
+	}
+	key := fmt.Sprintf("%s/%s:%v/ttl", TopologyInformationPath, is.info.IP, is.info.Port)
+	return util.PutKVToEtcd(ctx, is.etcdCli, keyOpDefaultRetryCnt, key,
+		time.Now().String(),
```
Comment (breeswish) [inline domain/infosync/info.go]: Is this a timestamp, or a human readable string?

Comment (mapleFU) [reply domain/infosync/info.go]: An RFC String

File: domain/domain.go (hunk 11 above)
Comment (mapleFU) [reply domain/domain.go]: Sorry, I didn't notice that. I add it now

File: domain/infosync/info.go (hunk 13 above)
Comment (breeswish) [reply domain/infosync/info.go]: I think using timestamp should be better so that you can check the aliveness easily at Dashboard side.

Comment (mapleFU) [reply domain/infosync/info.go]: Now it's timestamp

**Pull request merged**